sling/nlp/embedding/fact-plausibility.cc

```diff
@@ -0,0 +1,547 @@
+  void Checkpoint(int64 epoch, Network *model) override {
+    if (!model_filename_.empty()) {
+      LOG(INFO) << "Checkpoint model to " << model_filename_;
+      Save(*model, model_filename_);
+    }
+  }
+
+ private:
+  // Benchmark statistics.
+  struct Benchmark {
+    // Clear statistics.
+    void clear() {
+      positive_correct = positive_wrong = negative_correct = negative_wrong = 0;
+    }
+
+    // Add result to benchmark.
+    void add(bool golden, bool predicted) {
+      if (golden) {
+        if (predicted) {
+          positive_correct++;
+        } else {
+          positive_wrong++;
+        }
+      } else {
+        if (predicted) {
+          negative_wrong++;
+        } else {
+          negative_correct++;
+        }
+      }
+    }
+
+    // Add statistics from another benchmark.
+    void add(const Benchmark &other) {
+      positive_correct += other.positive_correct;
+      positive_wrong += other.positive_wrong;
+      negative_correct += other.negative_correct;
+      negative_wrong += other.negative_wrong;
+    }
+
+    // Total nummber of positive/negative examples.
+    float positive() const { return positive_correct + positive_wrong; }
+    float negative() const { return negative_correct + negative_wrong; }
+
+    // Accuracy for positive/negative examples.
+    float positive_accuracy() const { return positive_correct / positive(); }
+    float negative_accuracy() const { return negative_correct / negative(); }
+
+    // Number of positive/negative correct/wrong predictions.
+    int positive_correct = 0;
+    int positive_wrong = 0;
+    int negative_correct = 0;
+    int negative_wrong = 0;
+  };
+
+  // Evaluate model on positive and negative examples generated from instances.
+  void EvaluateModel(const Handles &instances, Benchmark *benchmark) {
+    static const int batch_size = 2;
+    Random rnd;
+    rnd.seed(seed_);
+    std::vector<Features> premises(batch_size);
+    std::vector<Features> hypotheses(batch_size);
+
+    Instance scorer(flow_.scorer);
+    int *premise = scorer.Get<int>(flow_.premise);
+    int *hypothesis = scorer.Get<int>(flow_.hypothesis);
+    float *logits = scorer.Get<float>(flow_.logits);
+
+    for (int n = 0; n < instances.size(); ++n) {
+      // Get instances for batch.
+      for (int i = 0; i < batch_size; ++i) {
+        int sample = rnd.UniformInt(instances.size());
+        Frame instance(&store_, instances[sample]);
+
+        // Split instance into premise and hypothesis.
+        int num_groups = instance.Get(p_groups_).AsArray().length();
+        int heldout = rnd.UniformInt(num_groups);
+        Split(instance, &premises[i], &hypotheses[i], heldout);
+      }
+
+      // Compute scores for all pairs in batch.
+      for (int i = 0; i < batch_size; ++i) {
+        for (int j = 0; j < batch_size; ++j) {
+          // Check that main hypothesis is not in premise for negative examples.
+          if (i != j && Has(premises[i], hypotheses[j][0])) continue;
```

Comment:
Split() truncates 'premises' and 'hypotheses' to max_features - 1, so Has() here can erroneously return false, since it could be looking for the hypothesis in a truncated list of premise facts. This may or may not be a big issue, but it is straightforward to change Split() not to do any truncation, so I would suggesting doing that. 
The truncation can be delegated to Copy().

Reply:
I thought it was easier to ensure that we newer overflowed the feature vector so I don't have to check for the case where there is no room for the terminating -1. You're right that we would miss a contradiction for a truncated item, but we have very few overflows, so I am not sure this matters.